ngrok/src/tunnel_ext.rs

```diff
@@ -33,98 +38,189 @@ use tokio::{
-) -> Result<(), io::Error>
-where
-    T: Tunnel + ?Sized,
-    F: FnMut(io::Error, Conn),
-{
-    let span = Span::current();
-    let path = addr.as_ref();
-    span.record("path", field::debug(&path));
-    loop {
-        trace!("waiting for new tunnel connection");
-        if !handle_one_pipe(this, path, &mut on_err).await? {
-            debug!("listener closed, exiting");
-            break;
+fn tls_config() -> Result<Arc<ClientConfig>, &'static io::Error> {
+    static CONFIG: Lazy<Result<Arc<ClientConfig>, io::Error>> = Lazy::new(|| {
+        let der_certs = rustls_native_certs::load_native_certs()?
+            .into_iter()
+            .map(|c| c.0)
+            .collect::<Vec<_>>();
+        let der_certs = der_certs.as_slice();
+        let mut root_store = RootCertStore::empty();
+        root_store.add_parsable_certificates(der_certs);
+        let config = ClientConfig::builder()
+            .with_safe_defaults()
+            .with_root_certificates(root_store)
+            .with_no_client_auth();
+        Ok(Arc::new(config))
+    });
+
+    Ok(CONFIG.as_ref()?.clone())
+}
+
+// Establish the connection to forward the tunnel stream to.
+// Takes the tunnel and connection to make additional decisions on how to wrap
+// the forwarded connection, i.e. reordering tls termination and proxyproto.
+// Note: this additional wrapping logic currently unimplemented.
+async fn connect<T: Tunnel + Send + ?Sized>(
+    _tunnel: &mut T,
+    _conn: &Conn,
+    url: &Url,
+) -> Result<Box<dyn IoStream>, io::Error> {
+    let host = url.host_str().unwrap_or("localhost");
+    Ok(match url.scheme() {
+        "tcp" => {
+            let port = url.port().ok_or(io::ErrorKind::InvalidInput)?;
+            let conn = connect_tcp(host, port).in_current_span().await?;
+            Box::new(conn)
+        }
+
+        "http" => {
+            let port = url.port().unwrap_or(80);
+            let conn = connect_tcp(host, port).in_current_span().await?;
+            Box::new(conn)
+        }
+
+        "https" | "tls" => {
+            let port = url.port().unwrap_or(443);
+            let conn = connect_tcp(host, port).in_current_span().await?;
+
+            // TODO: if the tunnel uses proxyproto, wrap conn here before terminating tls
+
+            let domain = rustls::ServerName::try_from(host)
+                .map_err(|_| io::Error::from(io::ErrorKind::InvalidInput))?;
+            Box::new(
+                async_rustls::TlsConnector::from(tls_config().map_err(|e| e.kind())?)
+                    .connect(domain, conn.compat())
+                    .await?
+                    .compat(),
+            )
+        }
+
+        #[cfg(not(target_os = "windows"))]
+        "unix" => Box::new(
+            UnixStream::connect(&format!(
+                "{}/{}",
+                url.host_str().unwrap_or_default(),
+                url.path()
+            ))
+            .await?,
+        ),
+
+        #[cfg(target_os = "windows")]
+        "pipe" => {
```

Comment:
Just so I'm understanding, this is supposed to look like `pipe://my-tunnel-socket-123456`?

Reply:
The normal pipe address looks like `\\.\pipe\ngrok_pipe`, it looks like you can get close with the substitutions in here, e.g.: `pipe:///pipe/ngrok_pipe` will get `\\.\\pipe\ngrok_pipe`. There's an extra backslash coming from the first forward slash in the `path` though, testing on my windows machine.

Reply:
OK, I dropped the substitutions entirely and now we support things like `pipe://hostname/pipename`, which turns into `\\hostname\pipe\pipename` or `pipe:pipename` which becomes `\\.\pipename`.